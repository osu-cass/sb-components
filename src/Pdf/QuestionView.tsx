import * as React from "react";
import { ItemView } from "./ItemView";
import { EvidenceStatement } from "./EvidenceStatement";
import { ItemCardTable } from "../ItemCard/ItemCardTable";
import { RubricTable } from "../Rubric/RubricTable";
import { QuestionModel } from "./PdfModels";
import { AboutItemModel } from "../AboutItem/AboutItemModels";
import { ScoringOptions } from "../Rubric/ScoringOptionsTable";

export interface QuestionViewProps {
  question: QuestionModel;
  displayScoreInfo: boolean;
}

export class QuestionView extends React.Component<QuestionViewProps, {}> {
  renderRubric(data?: AboutItemModel) {
    if (!this.props.displayScoreInfo) {
      return undefined;
    }
    let rubric: JSX.Element | undefined;
    if (data && data.sampleItemScoring && data.sampleItemScoring.rubrics) {
      const entries = data.sampleItemScoring.rubrics
        .map(r => r.rubricEntries)
        .reduce((acc, re) => acc.concat(re), []);
      const samples = data.sampleItemScoring.rubrics
        .map(r => r.samples)
        .reduce((acc, sample) => acc.concat(sample), []);

      if (entries.length > 1 || samples.length > 1) {
        rubric = <RubricTable rubrics={data.sampleItemScoring.rubrics} />;
      } else {
        const entriesJsx =
          samples.length > 0 ? (
            <div>
              <b>Rubric:</b> ({entries[0].scorepoint}point) {entries[0].value}
            </div>
          ) : (
            undefined
          );
        const samplesJsx =
          samples.length > 0 ? (
            <div>
              <b>Exemplar:</b> {samples[0]}
            </div>
          ) : (
            undefined
          );

        rubric = (
          <div>
            {samplesJsx} {entriesJsx}
          </div>
        );
      }
    }

    return rubric ? rubric : undefined;
  }

  render() {
    const data = this.props.question.data;

    const evidenceStatement =
      data && data.evidenceStatement ? (
        <EvidenceStatement statement={data.evidenceStatement} />
      ) : (
        undefined
      );

    const dataTable =
      data && data.itemCardViewModel ? (
        <ItemCardTable card={data.itemCardViewModel} />
      ) : (
        undefined
      );

    const scoringOptions =
      data &&
      data.sampleItemScoring &&
      data.sampleItemScoring.scoringOptions &&
      data.sampleItemScoring.scoringOptions.length > 0 ? (
        <ScoringOptions options={data.sampleItemScoring.scoringOptions} />
      ) : (
        undefined
      );

    const answerKey =
      data &&
      data.sampleItemScoring &&
      data.sampleItemScoring.answerKey &&
      this.props.displayScoreInfo ? (
        <div>
          <b>Key:</b> {data.sampleItemScoring.answerKey}
        </div>
      ) : (
        undefined
      );

    return (
      <div className="item">
        <p className="question-title">
          Question #{this.props.question.questionNumber}
        </p>
        {dataTable}
        {evidenceStatement}
        <ItemView view={this.props.question.view} />
        <hr />
        {answerKey}
        {this.renderRubric(data)}
        {scoringOptions}
      </div>
    );
  }
}
