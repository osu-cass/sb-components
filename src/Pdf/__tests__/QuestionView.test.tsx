import * as React from 'react';
import * as ReactDOM from 'react-dom';
import { QuestionView } from '../QuestionView';
import { shallow, mount, render } from 'enzyme';
import { questionDataNull } from './PdfMocks';


describe("QuestionDataTable", () => {
    it("matches snapshot", () => {
        expect(shallow(<QuestionView question={questionDataNull}/>)).toMatchSnapshot();
    })
})