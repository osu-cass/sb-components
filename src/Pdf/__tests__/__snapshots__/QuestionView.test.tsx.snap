// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QuestionDataTable matches snapshot 1`] = `
<div
  className="item"
>
  <p
    className="question-title"
  >
    Question #
    1
  </p>
  <ItemView
    view={
      Object {
        "captured": true,
        "html": undefined,
        "id": "12334",
        "picturePath": undefined,
        "type": 1,
      }
    }
  />
  <hr />
</div>
`;
